pkg/utils/machine/machine.go

```diff
@@ -15,18 +15,118 @@ limitations under the License.
 package machine
 
 import (
+	"context"
+	"errors"
+	"fmt"
+
 	"github.com/samber/lo"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
 	"github.com/aws/karpenter-core/pkg/operator/scheme"
 	"github.com/aws/karpenter-core/pkg/scheduling"
 )
 
+func EventHandler(ctx context.Context, c client.Client) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+		machine := o.(*v1alpha5.Machine)
+		nodeList := &v1.NodeList{}
+		if machine.Status.ProviderID == "" {
+			return nil
+		}
+		if err := c.List(ctx, nodeList, client.MatchingFields{"spec.providerID": machine.Status.ProviderID}); err != nil {
+			return nil
+		}
+		return lo.Map(nodeList.Items, func(n v1.Node, _ int) reconcile.Request {
+			return reconcile.Request{
+				NamespacedName: client.ObjectKeyFromObject(&n),
+			}
+		})
+	})
+}
+
+func NodeEventHandler(ctx context.Context, c client.Client) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+		node := o.(*v1.Node)
+		machineList := &v1alpha5.MachineList{}
+		if err := c.List(ctx, machineList, client.MatchingFields{"status.providerID": node.Spec.ProviderID}); err != nil {
+			return []reconcile.Request{}
+		}
+		return lo.Map(machineList.Items, func(m v1alpha5.Machine, _ int) reconcile.Request {
+			return reconcile.Request{
+				NamespacedName: client.ObjectKeyFromObject(&m),
+			}
+		})
+	})
+}
+
+type NodeNotFoundError struct {
+	ProviderID string
```

Comment:
Does this need to be exported if this is being accessed by the Error() function? Maybe it doesn't matter but we could do a `NewNodeNotFoundError(providerID string) NodeNotFoundError` and call it if you wanted to not export it.

same goes for the other errors in this file

Reply:
Generally, it's not best practice to have unexported types with exported fields. See [here](https://stackoverflow.com/questions/37012158/unexported-struct-best-practice-golint-complains)